ddl/ddl_tiflash_test.go

```diff
@@ -649,3 +661,122 @@ func TestAlterDatabaseBasic(t *testing.T) {
 	// There is less TiFlash store
 	tk.MustGetErrMsg("alter database tiflash_ddl set tiflash replica 3", "the tiflash replica count: 3 should be less than the total tiflash server count: 2")
 }
+
+func checkBatchPandingNum(t *testing.T, tkx *testkit.TestKit, level string, value string, ok bool) {
+	l := len(tkx.MustQuery(fmt.Sprintf("show %v variables where Variable_name='tidb_batch_pending_tiflash_count' and Value='%v'", level, value)).Rows())
+	if ok {
+		require.Equal(t, 1, l)
+	} else {
+		require.Equal(t, 0, l)
+	}
+}
+
+func TestTiFlashBatchAddVariables(t *testing.T) {
+	store, _, tear := createStoreWithoutMockTiFlash(t)
+	defer tear()
+
+	tk := testkit.NewTestKit(t, *store)
+	tk.MustExec("set SESSION tidb_batch_pending_tiflash_count=5")
+	tk.MustExec("set GLOBAL tidb_batch_pending_tiflash_count=6")
+
+	checkBatchPandingNum(t, tk, "session", "5", true)
+	checkBatchPandingNum(t, tk, "global", "6", true)
+	checkBatchPandingNum(t, tk, "global", "1.5", false)
+
+	tk.MustGetErrMsg("set GLOBAL tidb_batch_pending_tiflash_count=1.5", "[variable:1232]Incorrect argument type to variable 'tidb_batch_pending_tiflash_count'")
+	checkBatchPandingNum(t, tk, "global", "6", true)
+
+	tk2 := testkit.NewTestKit(t, *store)
+	checkBatchPandingNum(t, tk2, "session", "6", true)
+}
+
+func execWithTimeout(tk *testkit.TestKit, to time.Duration, sql string) (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), to)
+	defer cancel()
+	doneCh := make(chan error, 1)
+
+	go func(c context.Context) {
+		stmts, _ := tk.Session().Parse(context.Background(), sql)
+		for _, stmt := range stmts {
+			_, err := tk.Session().ExecuteStmt(ctx, stmt)
+			doneCh <- err
+			return
+		}
+		doneCh <- nil
+	}(ctx)
+	select {
+	case e := <-doneCh:
+		return false, e
+	case <-ctx.Done():
+		return true, nil
+	}
+}
+
+func TestTiFlashBasicRateLimiter(t *testing.T) {
+	s, teardown := createTiFlashContext(t)
+	defer teardown()
+	tk := testkit.NewTestKit(t, s.store)
+
+	threshold := 2
+	tk.MustExec("drop database if exists tiflash_ddl_limit")
+	tk.MustExec("create database tiflash_ddl_limit")
+	tk.MustExec(fmt.Sprintf("set SESSION tidb_batch_pending_tiflash_count=%v", threshold))
+	for i := 0; i < threshold; i++ {
+		tk.MustExec(fmt.Sprintf("create table tiflash_ddl_limit.t%v(z int)", i))
+	}
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/PollTiFlashReplicaStatusReplaceCurAvailableValue", `return(false)`))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/PollTiFlashReplicaStatusReplaceCurAvailableValue"))
+	}()
+
+	tk.MustExec(fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
+	tk.MustExec(fmt.Sprintf("create table tiflash_ddl_limit.t%v(z int)", threshold))
+	// The following statement shall fail, because it reaches limit
+	timeOut, err := execWithTimeout(tk, time.Second*3, fmt.Sprintf("alter database tiflash_ddl_limit set tiflash replica %v", 1))
```

Comment:
I don't quite understand why there is `execWithTimeout`. I mean it should fail even without timeout.

Reply:
I think it will success after `tiflashCheckPendingTablesRetry*tiflashCheckPendingTablesWaitTime`, which is quite a long time. I intend to do this, because I want to execute a DDL to update infoCache, if I think it may not be updated.

Reply:
I think it will be better if this DDL just stuck forever or fail, instead of success.

Reply:
The DDL stuck forever is not an intention and may cause goleak in test.
Meanwhile, if `succ+fail` reached threshold, we can just wait some time, then it will become below threshold again.